### ISSUE TITLE
Please add support for unique_ptr

### ISSUE THREAD

**Issue by ashwin** [swig/swig#692]
There is support for auto_ptr in v3.0.9, but no support yet for unique_ptr.

**Comment by myd7349** [comment]
Which target languages are you using SWIG for?

**Comment by ashwin** [comment]
Python

**Comment by AnushS4t1qbit** [comment]
This is definitely a popular feature. 

I am a little wary of volunteering to tackle this, since I am sure there is a good reason why this hasn't been implemented yet. 

If I were to go about making this happen though, where would be a good starting point, the implementation for `shared_ptr`?

**Comment by ojwb** [comment]
> I am sure there is a good reason why this hasn't been implemented yet

I doubt that reason is complexity though.  SWIG is developed by volunteers, and there are always more things that could be done than there is time to do them.

`std::auto_ptr` might be a better starting point, as `std::unique_ptr` is perhaps more akin to that.

**Comment by wsfulton** [comment]
See Lib/python/std_auto_ptr.i. As unique_ptr is so similar to auto_ptr,
there isn't much to change. However, there is only support for outputs, not
input. Probably the input ('in') typemap should check that the reference
count is exactly one and then take over ownership from python.

**Comment by lichti81** [comment]
I would be happy that there is also support for unique_ptr in C#

**Comment by KrisThielemans** [comment]
Are there any updates on this? We're also interested in Python. I realise that doing this an input seems quite hard. It might need a copy of the underlying object as Python etc users will not expect their objects to disappear I guess.

There's an interesting approach on [stackoverflow](https://stackoverflow.com/a/27699663/3467846)

**Comment by vadz** [comment]
FWIW #1722 adds support for `unique_ptr<>` in the "output" direction. I don't see much sense in doing it in the "input" direction but maybe I'm missing something.

**Comment by KrisThielemans** [comment]
Great news about the PR. Do you want to continue the discussion there?

My understanding of input vs output:
- output: `unique_ptr<Obj> create_obj();`
- input: `void use_obj(unique_ptr<Obj>);`

We're currently debating the latter on https://github.com/UCL/STIR/issues/485 where we try to find a safe way to let a class store a ptr. Our current thinking goes like this
```c++
class A
{
  private:
    shared_ptr<Obj> sptr;
  public:
    A(unique_ptr<Obj> p):  sptr(p)
    {}
   shared_ptr<const Obj> get_sptr() const { return sptr;}
};
```
In Python (or whatever wrapped language), we'd like to be able to do
```python
  obj = make_an_Obj();
  a = A(obj);
```
It'd be surprising if `obj` would no longer exist in Python (while in C++, the class interface tells you that's going  to happen).

**Comment by vadz** [comment]
> Great news about the PR. Do you want to continue the discussion there?

I think here is fine, I'll update the other PR with the changes requested by William when I can (adding `std_unique_ptr.i`) and it's probably more clear to keep the discussions there about that patch itself rather than more general questions.
 
> My understanding of input vs output:
> 
>     * output: `unique_ptr<Obj> create_obj();`
> 
>     * input: `void use_obj(unique_ptr<Obj>);`

Yes, I agree, this is how I understand it too.

> We're currently debating the latter on [UCL/STIR#485](https://github.com/UCL/STIR/issues/485) where we try to find a safe way to let a class store a ptr.

The only safe way is to clone the pointer. You can't take the object ownership across the language boundary.

> Our current thinking goes like this
> 
> ```c++
> class A
> {
>   private:
>     shared_ptr<Obj> sptr;
>   public:
>     A(unique_ptr<Obj> p):  sptr(p)
>     {}
>    shared_ptr<const Obj> get_sptr() const { return sptr;}
> };
> ```
> 
> In Python (or whatever wrapped language), we'd like to be able to do
> 
> ```python
>   obj = make_an_Obj();
>   a = A(obj);
> ```
> 
> It'd be surprising if `obj` would no longer exist in Python (while in C++, the class interface tells you that's going to happen).

I don't even think this would be possible, but if it were, I'm pretty sure it would be undesirable.

So cloning is the only safe solution, but I am not sure if it's appropriate for SWIG typemaps to perform such potentially expensive operation automatically.

**Comment by KrisThielemans** [comment]
> > We're currently debating the latter on UCL/STIR#485 where we try to find a safe way to let a class store a ptr.
>
> The only safe way is to clone the pointer. You can't take the object ownership across the language boundary.

agreed. I meant that we were discussing a safe way to do this on the C++ side only. It's only now that I've started to think about the implications for SWIG.

> > It'd be surprising if obj would no longer exist in Python (while in C++, the class interface tells you that's going to happen).
>
> I don't even think this would be possible, but if it were, I'm pretty sure it would be undesirable.

I suppose that you mean that you think it's impossible for the Python object to disappear, and that it'd be undesirable anyway. I certainly agree with that.

> So cloning is the only safe solution, but I am not sure if it's appropriate for SWIG typemaps to perform such potentially expensive operation automatically.

The question is though if there's an alternative. I can't see one. If I need to be able to call the `A(unique_ptr<Obj>)` constructor, how do I do it otherwise? Of course, I could create a hand-crafted extra function such as
```c++
shared_ptr<A> makeA(shared_ptr<const Obj> sp)
{ return make_shared<A>(make_unique(*sp)); }
```
and let SWIG wrap that. Or I could add this via `%extend`. But then I've essentially done a manual typemap I suppose.

Maybe you're saying that SWIG shouldn't be default have a typemap to do this, and the user need to enable this on a function/class basis. That'd be fine by me.

I unfortunately haven't ever looked into constructing my own typemaps, so wouldn't even know where to start (ok, I could read the SWIG doc!).

**Comment by lichti81** [comment]
Also important: It is possible to use `unique_ptr ` in containers, like this:
`void foo(std::vector<std::unique_ptr<MyClass>> value)`
`std::vector<std::unique_ptr<MyClass>> foo()`

SWIG typemap should also support something like this.

**Comment by wsfulton** [comment]
299880e6a658934d39ae3b53159a3b61f1051c16 which closes #1722 adds support for unique_ptr as an output. Note that move semantics are also recently supported by default for output types, see #1044. The implications for unique_ptr are that even if the new std_unique_ptr.i library file is not used, the wrapped code will now compile, even though the resulting code is not particularly useful as unique_ptr is treated as a pointer to unique_ptr, as is the case for any type that is not fully parsed by SWIG.

Handling move pointers like unique_ptr is an input is still something that is missing and I'll try give think about this next, including using the suggestions given here.

**Comment by wsfulton** [comment]
Regarding unique_ptr as in input, I have a proposed fix in 6736aab60ae978edc72c76b0c21c09d99a7398e5 which depends on e777b054d583aaf2ed445a263208c6ef21bc05ac. I hope the commit messages and testcases provide some useful information. This is just for Java atm. Please provide comments in the next few days otherwise, I'll extend to further languages. 

The generated code for wrapping
```c++
std::string takeKlassUniquePtr(std::unique_ptr<Klass> k) { ... }
```
in the cpp11_std_unique_ptr.i testcase is:
```c++
SWIGEXPORT jstring JNICALL Java_cpp11_1std_1unique_1ptr_cpp11_1std_1unique_1ptrJNI_takeKlassUniquePtr(JNIEnv *jenv, jclass jcls, jlong jarg1, jobject jarg1_) {
  jstring jresult = 0 ;
  std::unique_ptr< Klass > arg1 ;
  Klass *unique_temp1 ;
  std::string result;
  
  (void)jenv;
  (void)jcls;
  (void)jarg1_;
  unique_temp1 = *(Klass **)&jarg1;
  (&arg1)->reset(unique_temp1); 
  result = takeKlassUniquePtr(SWIG_STD_MOVE(arg1));
  jresult = jenv->NewStringUTF((&result)->c_str()); 
  return jresult;
}
```
where `SWIG_STD_MOVE` is simply `std::move` for C++11. You can see the pointer from the Java proxy class is owned by the unique_ptr and then moved into the wrapped function. The Java code is:
```java
  public static String takeKlassUniquePtr(Klass k) {
    return cpp11_std_unique_ptrJNI.takeKlassUniquePtr(Klass.swigRelease(k), k);
  }
```
where `Klass.swigRelease()` is used instead of `Klass.getCPtr()` as follows:
```java
  protected static long swigRelease(Klass obj) {
    long ptr = 0;
    if (obj != null) {
      if (!obj.swigCMemOwn)
        throw new RuntimeException("Cannot release ownership as memory is not owned");
      ptr = obj.swigCPtr;
      obj.swigCMemOwn = false;
      obj.delete();
    }
    return ptr;
  }
```
The Java code returns the underlying C++ pointer of the proxy class and then sets it to null, so same sort of behaviour as if the proxy class was a smart pointer doing what `std::unique_ptr` does when moved. It makes the proxy class no good for further use from Java, exactly what happens to a moved `std::unique_ptr` which also holds a null pointer after being moved. Subsequent behaviour of the proxy class is the same as if the `delete()` method was called on it. 

#999 is tracking move semantics generally, whereas this ticket is for the move-only `std::unique_ptr` type. These are closely related and a subsequent commit cd525631a0fd09178a2089f864c9ef3617b631b5 builds on the idea of the proxy class behaving like a smart pointer with move semantics. This commits handles rvalue references by assuming the function with rvalue references parameters will move the reference (as is the behaviour of move constructors and move assignment operators).

**Comment by christophe-calmejane** [comment]
Sorry to bump this one, but it looks like the PR doesn't handle unique_ptr with a custom deleter (second parameter of the unique_ptr template), am I correct?

**Comment by wsfulton** [comment]
Correct. Best raise a new issue for discussion on how this might be achieved.

**Comment by christophe-calmejane** [comment]
Thanks, done: #2411
